[NNC] Add Support For is_nan

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48975 [NNC] add fusion support for nan_to_num
* #48976 [NNC] Add fusion support of aten::to
* #48974 [NNC] add support for masked_fill
* **#48973 [NNC] Add Support For is_nan**

Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

Commit:
is_nan

[ghstack-poisoned]

Commit:
Update on "is_nan"




[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -952,6 +953,15 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
     } break;
 
+    case aten::isnan: {
+      return computeOneOperand("aten_relu", v, [](const ExprHandle& a) {
```
Maybe want to name this `aten_isnan` instead of `aten_relu`, although tbh it's been a TODO for a long time to give these silly things names corresponding to their IR values instead of their aten_op.

eellison replied:
oops lol, you can see where i started the implementation from

bertmaher commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -1321,11 +1323,32 @@ llvm::Value* LLVMCodeGenImpl::toVec(llvm::Value* v, int lanes) {
+    value_ = toVec(llvm::ConstantInt::get(dstType, 0), v->dtype().lanes());
+  } else {
+    TORCH_INTERNAL_ASSERT(v->dtype().scalar_type() == ScalarType::Int);
+    auto is_nan = irb_.CreateFCmpUNO(
```
Thanks to you my brain now shouts "UNO!!" whenever I see this :-D

bertmaher commented on torch/csrc/jit/tensorexpr/expr.cpp:
```diff
@@ -200,6 +200,10 @@ ExprHandle remainder(const ExprHandle& v1, const ExprHandle& v2) {
 }
 
+ExprHandle isnan(const ExprHandle& v1) {
+  return Intrinsics::make(kIsNan, v1);
```
qq: What's the dtype of this `Intrinsics::make` expr?  Does it default to whatever the type of `v1` is?  That would be fine here, but I wonder if we should make our intrinsics typing better (cc @penguinwu who is interested in that :) ).

eellison replied:
It is by default v1, but I added logic so that if we see kIsNan we return torch.bool (which is what we need)

bertmaher reviewed (approved):
Lgtm!

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add Support For is_nan"


Differential Revision: [D25413166](https://our.internmc.facebook.com/intern/diff/D25413166)

[ghstack-poisoned]
